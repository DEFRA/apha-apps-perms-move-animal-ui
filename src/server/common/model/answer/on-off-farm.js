import Joi from 'joi'
import { AnswerModel } from './answer-model.js'
import { validateAnswerAgainstSchema } from './validation.js'

const selectOptionText =
  'Select if you are moving cattle on or off your farm or premises'

export const onOffFarmPayloadSchema = Joi.object({
  onOffFarm: Joi.string().required().valid('on', 'off').messages({
    'any.required': selectOptionText,
    'any.valid': selectOptionText,
    'string.empty': selectOptionText
  })
})

/**
 * export @typedef {string} OnOffFarmData
 */

export class OnOffFarm extends AnswerModel {
  /**
   * @returns {OnOffFarmData}
   */
  toState() {
    return this._data?.onOffFarm ?? ''
  }

  get value() {
    return this._data?.onOffFarm
  }

  /**
   * @param {OnOffFarmData | undefined} state
   * @returns {OnOffFarm}
   */
  static fromState(state) {
    return new OnOffFarm(state !== undefined ? { onOffFarm: state } : {})
  }

  validate() {
    return validateAnswerAgainstSchema(onOffFarmPayloadSchema, this._data)
  }
}
