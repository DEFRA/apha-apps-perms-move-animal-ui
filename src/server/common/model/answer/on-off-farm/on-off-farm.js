import { RadioButtonAnswer } from '../radio-button/radio-button.js'
/** @import {RadioButtonConfig} from '../radio-button/radio-button.js' */

/** @type {RadioButtonConfig} */
const onOffFarmConfig = {
  payloadKey: 'onOffFarm',
  options: {
    on: { label: 'On to the farm or premises' },
    off: { label: 'Off the farm or premises' }
  },
  validation: {
    empty: 'Select if you are moving cattle on or off your farm or premises'
  }
}

/**
 * export @typedef {'on' | 'off'} OnOffFarmData
 * @typedef {{ onOffFarm: 'on' | 'off' }} OnOffFarmPayload
 */

/** @augments {RadioButtonAnswer<OnOffFarmPayload>} */
export class OnOffFarmAnswer extends RadioButtonAnswer {
  static config = onOffFarmConfig
}
