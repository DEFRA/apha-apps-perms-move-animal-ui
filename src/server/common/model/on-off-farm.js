import Joi from 'joi'
import { validateAgainstSchema, Model } from './model.js'
/** @import {RawPayload} from './model.js' */

const selectOptionText =
      'Select if you are moving cattle on or off your farm or premises'

export const onOffFarmPayloadSchema = Joi.object({
  onOffFarm: Joi.string().required().valid('on', 'off').messages({
    'any.required': selectOptionText,
    'any.valid': selectOptionText,
    'string.empty': selectOptionText
  })
})

/**
 * export @typedef {string} OnOffFarmData
 */

export class OnOffFarm extends Model {
  /** @type {RawPayload | undefined} */
  _data

  /**
   * @param {RawPayload | undefined} data
   */
  constructor(data) {
    super()
    this._data = data

    Object.seal(this) // makes class immutable after instantiation
  }

  /**
   * @returns {OnOffFarmData}
   */
  toState() {
    return this._data?.onOffFarm ?? ''
  }

  get value() {
    return {
      value: this._data?.onOffFarm
    }
  }

  /**
   * @param {OnOffFarmData | undefined} data
   * @returns {OnOffFarm}
   */
  static fromState(data) {
    return new OnOffFarm(data !== undefined ? { onOffFarm: data } : {})
  }

  validate() {
    return validateAgainstSchema(onOffFarmPayloadSchema, this._data)
  }
}

