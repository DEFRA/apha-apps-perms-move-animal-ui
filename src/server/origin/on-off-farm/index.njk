{% from "govuk/components/radios/macro.njk" import govukRadios %}
{% from "govuk/components/error-summary/macro.njk" import govukErrorSummary %}

{% extends 'layouts/questions.njk' %}

{% block questions %}
  {% if (errorMessages) %}
    {{ govukErrorSummary({
      titleText: "There is a problem",
      errorList: errorMessages
    }) }}
  {% endif %}

  {{ govukRadios({
    name: "onOffFarm",
    id: "onOffFarm",
    fieldset: {
      legend: {
        text: "Are you moving the cattle on or off your farm or premises?",
        isPageHeading: true,
        classes: "govuk-fieldset__legend--l"
      }
    },
    value: value,
    items: [
      {
        id: "on-farm-radio",
        value: "on",
        text: "On to the farm or premises"
      },
      {
        id: "off-farm-radio",
        value: "off",
        text: "Off the farm or premises"
      }
    ],
    errorMessage: errors.onOffFarm
  }) }}
{% endblock %}
