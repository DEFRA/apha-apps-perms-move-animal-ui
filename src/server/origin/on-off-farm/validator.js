import Joi from 'joi'

export const CphNumberSchema = Joi.string()
  .required()
  .valid('on', 'off')
  .messages({
    'any.required':
      'Select if you are moving cattle on or off your farm or premises',
    'any.valid':
      'Select if you are moving cattle on or off your farm or premises'
  })

/**
 * @param {string} cphNumber
 * @returns {{isValid: boolean, errors: object}}
 */
export default (cphNumber) => {
  const result = CphNumberSchema.options({ abortEarly: false }).validate(
    cphNumber
  )
  const errors = result.error?.details.map(({ message }) => [
    'cphNumber',
    { text: message }
  ])

  return {
    isValid: result.error === undefined,
    errors: errors && Object.fromEntries(errors)
  }
}
