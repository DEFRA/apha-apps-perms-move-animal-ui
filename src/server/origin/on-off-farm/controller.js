import { calculateNextPage } from '../../common/helpers/next-page.js'
import { OnOffFarm } from '~/src/server/common/model/answer/on-off-farm.js'

const pageTitle = 'Are you moving the cattle on or off your farm or premises?'
const heading = pageTitle
const indexView = 'origin/on-off-farm/index.njk'

/**
 * The first question to start the journey.
 * @satisfies {Partial<ServerRoute>}
 */
export const onOffFarmGetController = {
  handler(req, h) {
    const onOffFarm = OnOffFarm.fromState(req.yar.get('origin')?.onOffFarm)

    return h.view(indexView, {
      nextPage: req.query.redirect_uri,
      pageTitle,
      heading,
      onOffFarm
    })
  }
}

/**
 * Respond to the first question.
 * @satisfies {Partial<ServerRoute>}
 * @param req
 */
export const onOffFarmPostController = {
  handler(req, res) {
    const payload = /** @type {OnOffFarmPayload & NextPage} */ (req.payload)

    const onOffFarm = new OnOffFarm(payload)
    const { isValid, errors } = onOffFarm.validate()

    if (!isValid) {
      req.yar.set('origin', {
        ...req.yar.get('origin'),
        onOffFarm: undefined
      })

      return res.view(indexView, {
        nextPage: calculateNextPage(
          payload.nextPage,
          '/origin/to-or-from-own-premises'
        ),
        pageTitle: `Error: ${pageTitle}`,
        heading,
        errorMessage: errors.onOffFarm
      })
    }

    req.yar.set('origin', {
      ...req.yar.get('origin'),
      onOffFarm: onOffFarm.toState()
    })

    switch (payload.onOffFarm) {
      case 'off':
        return res.redirect(
          calculateNextPage(payload.nextPage, '/origin/cph-number')
        )
      case 'on':
        return res.redirect('/exit-page')
      default:
        return res.view(indexView, {
          pageTitle: `Error: ${pageTitle}`,
          heading,
          errorMessage: {
            text: 'Select if you are moving cattle on or off your farm or premises'
          }
        })
    }
  }
}

/**
 * @typedef {{ onOffFarm: 'on' | 'off' }} OnOffFarmPayload
 * @import { ServerRoute } from '@hapi/hapi'
 * @import {NextPage} from '../../common/helpers/next-page.js'
 */
