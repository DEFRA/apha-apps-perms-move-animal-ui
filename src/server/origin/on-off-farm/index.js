/**
 * Sets up the routes used in the on off farm page.
 * These routes are registered in src/server/router.js.
 */

import { OnOffFarm } from '~/src/server/common/model/answer/on-off-farm.js'
import { QuestionPage } from '../../common/model/page/question-page-model.js'
import { QuestionPageController } from '../../common/controller/question-page-controller.js'

/** @import { AnswerErrors } from "~/src/server/common/model/answer/validation.js" */
/** @import { AnswerModel } from "~/src/server/common/model/answer/answer-model.js" */

export class OnOffFarmPage extends QuestionPage {
  urlPath = '/origin/to-or-from-own-premises'
  sectionKey = 'origin'

  question = 'Are you moving the cattle on or off your farm or premises?'

  questionKey = 'onOffFarm'

  view = 'origin/on-off-farm/index'
  Answer = OnOffFarm

  /** @param {AnswerModel} answer */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  nextPage(answer) {
    return answer.nextPage()
  }

  /** @param {AnswerErrors} errors */
  errorMessages(errors) {
    return Object.entries(errors).map(([key, value]) => ({
      text: value.text,
      href: `#${key}`
    }))
  }
}

/**
 * @satisfies {ServerRegisterPluginObject<void>}
 */
export const onOffFarm = new QuestionPageController(
  new OnOffFarmPage()
).plugin()

/**
 * @import { ServerRegisterPluginObject } from '@hapi/hapi'
 */
