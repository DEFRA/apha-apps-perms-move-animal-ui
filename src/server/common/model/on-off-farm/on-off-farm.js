import Joi from 'joi'
import { validateAgainstSchema, Model } from '../model.js'
/** @import {RawPayload} from '../model.js' */

const selectOptionText =
  'Select if you are moving cattle on or off your farm or premises'

export const onOffFarmPayloadSchema = Joi.object({
  onOffFarm: Joi.string().required().valid('on', 'off').messages({
    'any.required': selectOptionText,
    'any.valid': selectOptionText,
    'string.empty': selectOptionText
  })
})

/**
 * export @typedef {string} OnOffFarmData
 */

export class OnOffFarm extends Model {
  /**
   * @returns {OnOffFarmData}
   */
  toState() {
    return this._data?.onOffFarm ?? ''
  }

  get value() {
    return this._data?.onOffFarm
  }

  /**
   * @param {OnOffFarmData | undefined} data
   * @returns {OnOffFarm}
   */
  static fromState(data) {
    return new OnOffFarm(data !== undefined ? { onOffFarm: data } : {})
  }

  validate() {
    return validateAgainstSchema(onOffFarmPayloadSchema, this._data)
  }
}
