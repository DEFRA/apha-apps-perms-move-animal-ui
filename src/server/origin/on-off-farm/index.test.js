import { onOffFarm, onOffFarmPage, OnOffFarmPage } from './index.js'
import { OnOffFarmAnswer } from '../../common/model/answer/on-off-farm/on-off-farm.js'
import { cphNumberPage } from '../cph-number/index.js'
import { exitPage } from '../../exit-page/index.js'

const sectionKey = 'origin'
const question = 'Are you moving the cattle on or off your farm or premises?'
const questionKey = 'onOffFarm'
const view = 'origin/on-off-farm/index'
const pageUrl = '/origin/to-or-from-own-premises'

describe('OnOffFarmPage', () => {
  let page

  beforeEach(() => {
    page = new OnOffFarmPage()
  })

  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct view', () => {
    expect(page.view).toBe(view)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(OnOffFarmAnswer)
  })

  it('nextPage should return exitPage when answer is "on"', () => {
    const answer = { value: 'on' }
    const nextPage = page.nextPage(answer)
    expect(nextPage).toBe(exitPage)
  })

  it('nextPage should return cphNumberPage when answer is "off"', () => {
    const answer = { value: 'off' }
    const nextPage = page.nextPage(answer)
    expect(nextPage).toBe(cphNumberPage)
  })

  it('should export page', () => {
    expect(onOffFarmPage).toBeInstanceOf(OnOffFarmPage)
  })

  it('should export emailAddress as a plugin', () => {
    expect(onOffFarm).toHaveProperty('plugin')
    const plugin = /** @type {PluginBase<void> & PluginNameVersion} */ (
      onOffFarm.plugin
    )
    expect(plugin).toHaveProperty('name')
    expect(plugin.name).toBe(`${sectionKey}-${questionKey}`)
    expect(plugin).toHaveProperty('register')
  })
})

/** @import { PluginBase, PluginNameVersion } from '@hapi/hapi' */
