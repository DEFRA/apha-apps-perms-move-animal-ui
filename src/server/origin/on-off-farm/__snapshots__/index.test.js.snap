// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnOffFarmPage #onOffFarmPage.content licenceSummaryPage.content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset">
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
      <h1 class="govuk-fieldset__heading">
        Are you moving the animals on or off your farm or premises?
      </h1>
    </legend>
    <div class="govuk-radios" data-module="govuk-radios">
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="onOffFarm" name="onOffFarm" type="radio" value="on">
        <label class="govuk-label govuk-radios__label" for="onOffFarm">
          On to the farm or premises
        </label>
      </div>
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="off" name="onOffFarm" type="radio" value="off">
        <label class="govuk-label govuk-radios__label" for="off">
          Off the farm or premises
        </label>
      </div>
    </div>
  </fieldset>
</div>

            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
